File: julia/benchmark_drazin.jl
# Headline experiment (Julia): on NON-symmetric singular systems with controlled
# Drazin index, plain GMRES (dgmres index=0) converges to the WRONG solution,
# while DGMRES (index=k) recovers the exact Drazin solution A^D b.
#
# Honest metrics reported:
#   res    = ‖A x - b‖ / ‖b‖         (ordinary residual — GMRES minimises THIS)
#   derr   = ‖x - A^D b‖ / ‖A^D b‖   (error vs the true Drazin solution)
# GMRES wins on `res` yet loses on `derr`: it solves a different problem.
#
# Run:  julia code/julia/benchmark_drazin.jl
include("DrazinKrylov.jl")
using .DrazinKrylov
using .DrazinKrylov: drazin_solution_block
using LinearAlgebra, Printf, Random

const SIZES = (1000, 2000, 4000)
const INDICES = (1, 2, 3)
const CAP_M = 400

function run_case(n_core, k)
    A, coreblk, nilblk = sparse_block_singular(n_core, k; seed = 12345)
    n = n_core + k
    b = randn(MersenneTwister(999), n)
    xD = drazin_solution_block(A, b, coreblk, nilblk)
    nrmb = norm(b); nrmxD = max(norm(xD), 1e-30)

    results = Dict{String,Any}()
    for (name, idx) in (("GMRES", 0), ("DGMRES", k))
        t = @elapsed res = dgmres(A, b; index = idx, m = min(n, CAP_M), tol = 1e-10)
        x = res.x
        rr = norm(A * x - b) / nrmb
        de = norm(x - xD) / nrmxD
        results[name] = (t = t, iters = res.iterations, conv = res.converged, res = rr, derr = de)
    end
    return results
end

# --- JIT warmup (compile everything on a tiny problem; NOT timed) ---
let
    A, cb, nb = sparse_block_singular(20, 2; seed = 1)
    b = randn(MersenneTwister(1), 22)
    dgmres(A, b; index = 0, m = 22, tol = 1e-10)
    dgmres(A, b; index = 2, m = 22, tol = 1e-10)
    drazin_solution_block(A, b, cb, nb)
end

open(joinpath(@__DIR__, "..", "results", "drazin_julia.csv"), "w") do io
    println(io, "lang,n,index,method,time_s,iters,converged,rel_residual,drazin_error")
    @printf("%-6s %-7s %-4s %-7s %10s %6s %5s %12s %12s\n",
            "lang", "n", "k", "method", "time_s", "iters", "conv", "rel_res", "drazin_err")
    for n_core in SIZES, k in INDICES
        r = run_case(n_core, k)
        for method in ("GMRES", "DGMRES")
            m = r[method]
            @printf("%-6s %-7d %-4d %-7s %10.4f %6d %5s %12.3e %12.3e\n",
                    "Julia", n_core + k, k, method, m.t, m.iters, m.conv, m.res, m.derr)
            @printf(io, "Julia,%d,%d,%s,%.6f,%d,%s,%.6e,%.6e\n",
                    n_core + k, k, method, m.t, m.iters, m.conv, m.res, m.derr)
        end
    end
end
println("\nwrote code/results/drazin_julia.csv")

File: python/benchmark_drazin.py
# Headline experiment (Python mirror of benchmark_drazin.jl).
# GMRES (dgmres index=0) converges to the WRONG solution on non-symmetric
# singular systems; DGMRES (index=k) recovers the exact Drazin solution A^D b.
# Run:  python code/python/benchmark_drazin.py
import os
import time
import numpy as np
import scipy.sparse.linalg as spla
from drazin_krylov import sparse_block_singular, dgmres

SIZES = (1000, 2000, 4000)
INDICES = (1, 2, 3)
CAP_M = 400
RESULTS = os.path.join(os.path.dirname(__file__), "..", "results", "drazin_python.csv")


def drazin_solution_block(A, b, n_core):
    """Closed-form A^D b for sparse_block_singular: x[core]=B^-1 b[core], x[nil]=0."""
    x = np.zeros_like(b)
    B = A[:n_core, :n_core].tocsc()
    x[:n_core] = spla.spsolve(B, b[:n_core])
    return x


def run_case(n_core, k):
    A, _, _ = sparse_block_singular(n_core, k, seed=12345)
    n = n_core + k
    b = np.random.default_rng(999).standard_normal(n)
    xD = drazin_solution_block(A, b, n_core)
    nrmb = np.linalg.norm(b)
    nrmxD = max(np.linalg.norm(xD), 1e-30)
    out = {}
    for name, idx in (("GMRES", 0), ("DGMRES", k)):
        t0 = time.perf_counter()
        res = dgmres(A, b, index=idx, m=min(n, CAP_M), tol=1e-10)
        t = time.perf_counter() - t0
        x = res.x
        rr = np.linalg.norm(A @ x - b) / nrmb
        de = np.linalg.norm(x - xD) / nrmxD
        out[name] = dict(t=t, iters=res.iterations, conv=res.converged, res=rr, derr=de)
    return out


# JIT/import warmup so timing isn't polluted by first-call overhead
_A, _, _ = sparse_block_singular(20, 2, seed=1)
_b = np.random.default_rng(1).standard_normal(22)
dgmres(_A, _b, index=0, m=22, tol=1e-10)
dgmres(_A, _b, index=2, m=22, tol=1e-10)

hdr = f'{"lang":<7}{"n":<7}{"k":<4}{"method":<8}{"time_s":>10}{"iters":>7}{"conv":>6}{"rel_res":>12}{"drazin_err":>13}'
print(hdr)
with open(RESULTS, "w") as f:
    f.write("lang,n,index,method,time_s,iters,converged,rel_residual,drazin_error\n")
    for n_core in SIZES:
        for k in INDICES:
            r = run_case(n_core, k)
            for method in ("GMRES", "DGMRES"):
                m = r[method]
                print(f'{"Python":<7}{n_core + k:<7}{k:<4}{method:<8}{m["t"]:>10.4f}'
                      f'{m["iters"]:>7}{str(m["conv"]):>6}{m["res"]:>12.3e}{m["derr"]:>13.3e}')
                f.write(f'Python,{n_core + k},{k},{method},{m["t"]:.6f},{m["iters"]},'
                        f'{m["conv"]},{m["res"]:.6e},{m["derr"]:.6e}\n')
print("\nwrote code/results/drazin_python.csv")
